File: bms/work.py
from collections import defaultdict
from typing import List, Tuple


def extract_work_name(
    titles: List[str],
    remove_unclosed_pair: bool = True,
    remove_tailing_sign_list: List[str] = [],
) -> str:
    """
    从多个BMS文件标题中提取共同的作品名（改进版）

    :param titles: 包含多个BMS标题的列表
    :return: 提取出的共同作品名（经过后处理）
    """
    # 统计所有可能前缀的出现次数
    prefix_counts = defaultdict(int)
    for title in titles:
        for i in range(1, len(title) + 1):
            prefix = title[:i]
            prefix_counts[prefix] += 1

    if not prefix_counts:
        return ""

    max_count = max(prefix_counts.values())

    candidates = [
        (prefix, count)
        for prefix, count in prefix_counts.items()
        if count >= max_count * 0.67  # 超过2/3就可以算上
    ]

    # 排序规则：优先长度降序，其次次数降序，最后字典序升序
    candidates.sort(key=lambda x: (-len(x[0]), -x[1], x[0]))

    # 提取最优候选
    best_candidate = candidates[0][0] if candidates else ""

    # 后处理：移除未闭合括号及其后续内容
    return _extract_work_name_post_process(
        best_candidate,
        remove_unclosed_pair=remove_unclosed_pair,
        remove_tailing_sign_list=remove_tailing_sign_list,
    )


def _extract_work_name_post_process(
    s: str, remove_unclosed_pair: bool = True, remove_tailing_sign_list: List[str] = []
) -> str:
    """
    后处理函数：移除未闭合括号及其后续内容

    :param s: 原始字符串
    :return: 处理后的字符串
    """

    # 清除前后空格
    s = s.strip()

    while True:
        triggered = False
        if remove_unclosed_pair:
            stack: List[Tuple[str, int]] = []

            # 遍历字符串记录括号状态
            pairs = [
                ("(", ")"),
                ("[", "]"),
                ("{", "}"),
                ("（", "）"),
                ("［", "］"),
                ("｛", "｝"),
                ("【", "】"),
            ]
            for i, c in enumerate(s):
                for p_open, p_close in pairs:
                    if c == p_open:
                        stack.append((c, i))  # 记录括号类型和位置
                    if c == p_close and stack and stack[-1][0] == p_open:
                        stack.pop()

            # 如果存在未闭合括号
            if stack:
                last_unmatched_pos = stack[-1][1]
                s = s[:last_unmatched_pos].rstrip()  # 截断并移除末尾空格
                triggered = True

        for sign in remove_tailing_sign_list:
            if s.endswith(sign):
                s = s[: -len(sign)].rstrip()
                triggered = True
        # 没触发？
        if not triggered:
            break

    return s

File: be-music-cabinet/src/bms/work.rs
use std::cmp::Ordering;
use std::collections::HashMap;

/// 从多个标题中提取共同的作品名（改进版）
///
/// # 参数
/// - `titles`: 包含多个标题的切片
/// - `remove_unclosed_pair`: 是否移除未闭合括号对（默认为 true）
/// - `remove_tailing_sign_list`: 需要移除的尾部符号列表（默认为空）
///
/// # 返回值
/// 提取出的共同作品名（经过后处理）
pub fn extract_work_name(
    titles: &[&str],
    remove_unclosed_pair: bool,
    remove_tailing_sign_list: &[&str],
) -> String {
    // 如果标题列表为空，直接返回空字符串
    if titles.is_empty() {
        return "[!!! EMPTY !!!]".to_string();
    }

    // 使用 HashMap 统计所有可能前缀的出现次数
    let mut prefix_counts: HashMap<String, usize> = HashMap::new();

    // 遍历每个标题
    for &title in titles {
        // 生成标题的所有前缀（按字符长度）
        for i in 1..=title.chars().count() {
            // 获取前 i 个字符组成的前缀
            let prefix: String = title.chars().take(i).collect();
            // 更新前缀计数
            *prefix_counts.entry(prefix).or_insert(0) += 1;
        }
    }

    // 找出最大出现次数
    let max_count = *prefix_counts.values().max().unwrap_or(&0);

    // 筛选候选前缀：出现次数超过最大次数的 2/3
    let mut candidates: Vec<(String, usize)> = prefix_counts
        .into_iter()
        .filter(|(_, count)| *count as f32 >= max_count as f32 * 0.67)
        .collect();

    // 对候选前缀排序（优先长度降序，其次次数降序，最后字典序升序）
    candidates.sort_by(|a, b| {
        // 1. 按长度降序排序
        let len_cmp = b.0.len().cmp(&a.0.len());
        if len_cmp != Ordering::Equal {
            return len_cmp;
        }

        // 2. 按出现次数降序排序
        let count_cmp = b.1.cmp(&a.1);
        if count_cmp != Ordering::Equal {
            return count_cmp;
        }

        // 3. 按字典序升序排序
        a.0.cmp(&b.0)
    });

    // 提取最优候选（若存在则取第一个，否则为空字符串）
    let best_candidate = candidates
        .first()
        .map(|(s, _)| s.clone())
        .unwrap_or_default();

    // 对最优候选进行后处理
    extract_work_name_post_process(
        &best_candidate,
        remove_unclosed_pair,
        remove_tailing_sign_list,
    )
}

/// 作品名后处理函数：移除未闭合括号和尾部符号
///
/// # 参数
/// - `s`: 原始字符串
/// - `remove_unclosed_pair`: 是否处理未闭合括号
/// - `remove_tailing_sign_list`: 需要移除的尾部符号列表
///
/// # 返回值
/// 处理后的字符串
fn extract_work_name_post_process(
    s: &str,
    remove_unclosed_pair: bool,
    remove_tailing_sign_list: &[&str],
) -> String {
    // 先去除首尾空白字符
    let mut result = s.trim().to_string();

    // 定义支持的括号对（包含全角和半角）
    const PAIRS: [(char, char); 7] = [
        ('(', ')'),
        ('[', ']'),
        ('{', '}'),
        ('（', '）'),
        ('［', '］'),
        ('｛', '｝'),
        ('【', '】'),
    ];

    // 循环处理直到没有变化
    loop {
        let mut changed = false;

        // 处理未闭合括号
        if remove_unclosed_pair {
            // 存储括号状态（括号字符 + 字节位置）
            let mut stack: Vec<(char, usize)> = Vec::new();

            // 遍历字符串的每个字符（带字节索引）
            for (byte_idx, c) in result.char_indices() {
                // 检查是否是开括号
                if PAIRS.iter().any(|&(open, _)| c == open) {
                    stack.push((c, byte_idx));
                }
                // 检查是否是闭括号
                else if let Some(&(last_open, _)) = stack.last() {
                    // 查找匹配的闭括号
                    if let Some(&(_, close)) = PAIRS.iter().find(|&&(open, _)| open == last_open)
                        && c == close
                    {
                        stack.pop();
                    }
                }
            }

            // 如果存在未闭合括号
            if let Some(&(_, unmatched_pos)) = stack.last() {
                // 截断到第一个未匹配括号的位置
                result.truncate(unmatched_pos);
                // 移除截断后的尾部空白
                result = result.trim_end().to_string();
                changed = true;
            }
        }

        // 处理尾部符号
        for &sign in remove_tailing_sign_list {
            if result.ends_with(sign) {
                // 移除匹配的尾部符号
                result.truncate(result.len() - sign.len());
                // 移除尾部空白
                result = result.trim_end().to_string();
                changed = true;
                // 一次只移除一个符号，然后重新检查
                break;
            }
        }

        // 如果没有变化则结束处理
        if !changed {
            break;
        }
    }

    result
}
